### PR TITLE
:sparkles: Game library

### PR DESCRIPTION
**Pull request by BigBrainAFK**
New stuff:
- Game library in main window
- Moved the `Start`, `Pause`, `Restart` and `Screenshot` button to a menu bar item called `Emulator`
- Some plastic surgery on the UI files cause tabs are better than spaces

### CODE REVIEW

File: mainwindow.ui [hunk 1]
```diff
@@ -1,110 +1,111 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <ui version="4.0">
- <class>MainWindow</class>
```
Comment (mborgerson) [inline mainwindow.ui]: Why is this massive re-format necessary?... it creates a lot of churn and makes it harder to review the actual changes. Also, if a change like this must be done, I think it should at least be broken into a separate commit. Also also, this file was created and is typically edited using the QT interface designer tool.

File: main.py [hunk 2]
```diff
@@ -344,53 +374,95 @@ def __init__(self, *args):
...
+		self.fileformats = [
+			{
+				'name': 'ISO9660 CD/DVD image file',
+				'offset': [
+					0x8001,
+					0x8801,
+					0x9001
+				],
+				'signature': [
+					'43 44 30 30 31'
```
Comment (mborgerson) [inline main.py]: Why store it like this and not just 'CD001'?

File: main.py [hunk 3]
```diff
@@ -344,53 +374,95 @@ def __init__(self, *args):
...
+				]
+			},
+			{
+				'name': 'XISO',
+				'offset': [
+					0x10000
+				],
+				'signature': [
+					" ".join("{:02x}".format(ord(c)) for c in 'MICROSOFT*XBOX*MEDIA')
```
Comment (mborgerson) [inline main.py]: Same comment as above

**Comment by mborgerson** [comment]
I've paused my review of this PR for the moment because there are multiple changes in this PR which I think merit individual PRs. Can you please divide this PR into multiple PRs so we can review them independently? In my opinion, changes like moving the "Start" button to a menu item, adding a game library, and re-formatting the entire UI file, should be separate commits (which they are not here), and separate PRs--which we can then discuss and review individually.

File: main.py (hunk 2 above)
Comment (BigBrainAFK) [reply main.py]: because it gets compared to bytes

File: main.py (hunk 3 above)
Comment (BigBrainAFK) [reply main.py]: same as above

File: main.py (hunk 2 above)
Comment (JayFoxRox) [reply main.py]: b'CD001'?

File: main.py (hunk 3 above)
Comment (JayFoxRox) [reply main.py]: This is a unicode string. `'MICROSOFT*XBOX*MEDIA'.encode('utf-16')` or something will do the trick.
(Microsoft / Xbox used UCS-2 encoding normally)

File: main.py [hunk 4]
```diff
@@ -2,23 +2,38 @@
...
 
 SETTINGS_FILE = './settings.json'
 
 # Load UI files
 settings_class, _ = loadUiType('settings.ui')
 mainwindow_class, _ = loadUiType('mainwindow.ui')
 
+def read_from_hex_offset(file, hex_offset):
+	file.seek(hex_offset)
+	return file.read(32)
```
Comment (JayFoxRox) [inline main.py]: This function is a bit... odd? It feels unnecessary, and a bit "stiff".

* What's a "hex_offset"?
* Might want to be able to specify size, too (why 32 bytes?).

File: main.py [hunk 5]
```diff
@@ -2,23 +2,38 @@
...
+
+def getFormat(fileformats, settings, filename):
+	for element in fileformats:
+		for offset in element['offset']:
+			file = open(settings.settings['iso_path'] + '/' + filename, 'rb')
```
Comment (JayFoxRox) [inline main.py]: I suggest to give this function an absolute path.

So if we allow an "open-from-file" from outside of the ISO path, we can still use it.

File: main.py [hunk 6]
```diff
@@ -2,23 +2,38 @@
...
 settings_class, _ = loadUiType('settings.ui')
 mainwindow_class, _ = loadUiType('mainwindow.ui')
 
+def read_from_hex_offset(file, hex_offset):
+	file.seek(hex_offset)
+	return file.read(32)
+
+def getFormat(fileformats, settings, filename):
+	for element in fileformats:
+		for offset in element['offset']:
```
Comment (JayFoxRox) [inline main.py]: The way this loop is written, the file is opened/closed for every checked format.
You should open the file once, and then check for all formats at the same time.

File: main.py [hunk 7]
```diff
@@ -2,23 +2,38 @@
...
+			file = open(settings.settings['iso_path'] + '/' + filename, 'rb')
+			hex_bytes = " ".join("{:02x}".format(ord(c)) for c in read_from_hex_offset(file, offset))
```
Comment (JayFoxRox) [inline main.py]: This seems overly complicated. We should never require to work with hex-strings.
Instance shortly after, too.

File: main.py [hunk 8]
```diff
@@ -32,6 +47,7 @@ def reset(self):
 			'hdd_locked': True,
 			'dvd_present': True,
 			'dvd_path': '/path/to/disc.iso',
+			'iso_path': '/path/to/iso_folder',
```
Comment (JayFoxRox) [inline main.py]: We not only support ISOs, I'd say `drive_path` or `disc_path` or something.
We can also get rid of the `dvd_path` at the same time probably?

File: main.py (hunk 6 above)
Comment (BigBrainAFK) [reply main.py]: that i do have to say is a good point will change that

File: main.py [hunk 9]
```diff
@@ -148,6 +181,15 @@ def setSaveFileName(self, obj):
 		if fileName:
 			obj.setText(fileName)
 
+	def setDirectory(self, obj):
+		options = QFileDialog.Options()
+		directory = QFileDialog.getExistingDirectory(self,
+				"Select ISO Directory",
```
Comment (JayFoxRox) [inline main.py]: Same issue as mentioned before: We not only support ISOs (in fact, the original MS bios does not support ISO[9660] by default as far as I know).

File: main.py [hunk 10]
```diff
@@ -232,9 +274,10 @@ def escape_path(path):
 		accelerator_arg = Xqemu.generateAcceleratorArg(settings.settings['use_accelerator'])
 
 		dvd_path_arg = ''
-		if settings.settings['dvd_present']:
-			check_path(settings.settings['dvd_path'])
-			dvd_path_arg = ',file=' + escape_path(settings.settings['dvd_path'])
+		if mwindow.tableGames.currentRow() > 0:
+			selectedGame = settings.settings['iso_path'] + '/' + mwindow.tableGames.item(mwindow.tableGames.currentRow(), 0).text().split('\n')[1]
```
Comment (JayFoxRox) [inline main.py]: The filename should not be extracted from the UI element.
We should store extra-data per element, so the UI and logic are entirely seperated.

File: main.py [hunk 11]
```diff
@@ -274,8 +317,8 @@ def launchCmdToString(cmd):
 
 		return ' '.join(cmd_escaped)
 
-	def start(self, settings):
-		cmd = self.generateLaunchCmd(settings)
+	def start(self, settings, mwindow):
+		cmd = self.generateLaunchCmd(settings, mwindow)
```
Comment (JayFoxRox) [inline main.py]: What's the "m" in `mwindow`?

File: main.py [hunk 12]
```diff
@@ -344,53 +387,95 @@ def __init__(self, *args):
...
+				'offset': [
+					0x9001,
+					0x8801,
+					0x8001
+				],
+				'signature': [
+					'CD001'
+				]
+			}
+		]
```
Comment (JayFoxRox) [inline main.py]: I think you can hard-code this (in a separate function). It's unlikely we'll add more formats, and even if we do, it's only like 5 lines per format.

We might also want to perform additional checks for some formats, and adding this here would be.. complicated and weird?

File: settings.ui [hunk 13]
```diff
@@ -89,20 +89,13 @@
        <item>
         <widget class="QGroupBox" name="dvdDriveGroup">
          <property name="title">
-          <string>DVD Drive</string>
+          <string>ISO Folder</string>
```
Comment (JayFoxRox) [inline settings.ui]: "ISO": Same as other instances.
